### PR TITLE
feat: Auto-zoom keymap layout

### PR DESCRIPTION
**Pull request by Nicell**
Adds auto scaling the keymap/layout to fit available space. Adds a selection in the top right to override auto to a specific percentage zoom, too.

image

### CODE REVIEW

File: src/keyboard/Keyboard.tsx [hunk 1]
```diff
@@ -514,15 +516,37 @@ export default function Keyboard() {
...
           <KeymapComp
             keymap={keymap}
             layout={layouts[selectedPhysicalLayoutIndex]}
             behaviors={behaviors}
+            scale={keymapScale}
             selectedLayerIndex={selectedLayerIndex}
             selectedKeyPosition={selectedKeyPosition}
             onKeyPositionClicked={setSelectedKeyPosition}
           />
+          <select
```
Comment (petejohanson) [inline src/keyboard/Keyboard.tsx:533]: Do we want to use a bare `select` here or used a styled React Aria [Select](https://react-spectrum.adobe.com/react-aria/Select.html) for this?

**Review by petejohanson — COMMENTED** [review]
One code comment, and a general feature thought: Should we persist the selected zoom level to local user settings and restore it for future loads of the app?

File: src/keyboard/Keyboard.tsx (hunk 1 above)
Comment (Nicell) [reply src/keyboard/Keyboard.tsx:533]: I was following along with the general usage of `select` over the React Aria one. I tried the React Aria one and didn't like the initial results, so if there are future plans to move everything to React Aria's version, there should probably be some time put into a wrapped select with common styling.

**Comment by Nicell** [comment]
Added local storage persistence 👍

File: src/keyboard/Keyboard.tsx (hunk 1 above)
Comment (petejohanson) [reply src/keyboard/Keyboard.tsx:533]: Fair. WFM.

File: src/keyboard/Keyboard.tsx [hunk 2]
```diff
@@ -172,6 +173,15 @@ export default function Keyboard() {
...
+  const [keymapScale, setKeymapScale] = useState<LayoutZoom>(() => {
+    const savedValue = localStorage.getItem("keymapScale");
+    return savedValue === null || savedValue === "auto" ? "auto" : parseFloat(savedValue) || "auto";
+  });
+
+  useEffect(() => {
+    localStorage.setItem("keymapScale", keymapScale.toString());
+  }, [keymapScale]);
+
   const [selectedLayerIndex, setSelectedLayerIndex] = useState<number>(0);
```
Comment (petejohanson) [inline src/keyboard/Keyboard.tsx]: Given I'm sure we'll have other localStorage state down the road, can we pull this little bit of isolated logic into some new `useLocalStorageState` hook?

Comment (Nicell) [reply src/keyboard/Keyboard.tsx]: Done!

File: src/misc/useLocalStorageState.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    return defaultValue;
+  })
+
+  const [state] = reactState;
+
+  useEffect(() => {
+    const serializedState = options?.serialize ? options.serialize(state)
+      : (typeof state === 'object' && state !== null)
+        ? JSON.stringify(state) // Fallback for objects
+        : String(state); // Fallback for other types
```
Comment (petejohanson) [inline src/misc/useLocalStorageState.ts]: Don't love the nested tertiary here... Can we at least simplify it along the lines of:

```suggestion
    const serializedState = options?.serialize?(state) || 
      ((typeof state === 'object' && state !== null)
        ? JSON.stringify(state) // Fallback for objects
        : String(state)); // Fallback for other types
```

Or if not, at least:

```suggestion
    const serializedState = options?.serialize ? options.serialize(state) : basicSerialization(state);
```

And move the nested code into a small helper function?

**Review by petejohanson — CHANGES_REQUESTED** [review]

**Review by petejohanson — APPROVED** [review]
Thanks!

**Pull request merged**